### DISCUSSION TITLE
One Script to Rule them All

### DISCUSSION THREAD

**Discussion by ADavidBailey** [ADavidBailey/Practice-Bidding-Scenarios#46] [General]
I starting to put together [one cmd file that will update everything](https://github.com/ADavidBailey/Practice-Bidding-Scenarios/blob/main/build-scripts/oneScriptToRuleThemAll).  I intend to run it on Windows.  Needless to say, I'll need your help with this.  Thanks in advance

@thorvaldaagaard @rick-wilson @edwardpiwowar

**Comment by ADavidBailey** [comment]
I have a question, already.  I created a testing.cmd that just has two statements:

python3 extract.py              # /PBS > /dlr
./makeFiltered out              # /bba > /bba-filtered-out

If/when these two run, they won't change anything.  I'm using them to figure out how to get things to work.  When I open a PowerShell, go to the P:\build-scripts and enter ./testing.cmd, I get this:

PS P:\build-scripts> ./testing.cmd

P:\build-scripts>python3 extract.py              # /PBS  1>/dlr
Access is denied.

P:\build-scripts>./makeFiltered out              # /bba  1>/bba-filtered-out
Access is denied.
PS P:\build-scripts>

After googling, I changed the execution policy

PS P:\build-scripts> Get-ExecutionPolicy -list

        Scope ExecutionPolicy
        ----- ---------------
MachinePolicy       Undefined
   UserPolicy       Undefined
      Process       Undefined
  CurrentUser    RemoteSigned
 LocalMachine    RemoteSigned

Still no joy!  What should I do?

**Reply by ADavidBailey** [reply]
Do I have to run VSCode on Windows?

**Reply by ThorvaldAagaard** [reply]
Can it be / - Windows default use \
And I would use Visual Code but don't think it will change anything about the execution

**Reply by ADavidBailey** [reply]
> Can it be / - Windows default use

The commands work when they're NOT in the PowerShell cmd.

**Reply by ADavidBailey** [reply]
> If/when these two run, they won't change anything

they don't do anything, much -- does anyone care about the 'Created'?  It's nice that GitHub lets me delete these 221 updates with a single click.
![Screenshot 2024-10-24 at 6 52 49 AM](https://github.com/user-attachments/assets/1fcde11d-c3ed-4f1b-8264-5bbf11191f03)

**Reply by ThorvaldAagaard** [reply]
I asked ChatGPT
![image](https://github.com/user-attachments/assets/3faee563-9c42-485b-aec7-6d9878161a63)
![image](https://github.com/user-attachments/assets/809ee524-3bfb-40ba-97c4-4c4acae91eb6)

Seems there is some commands there you should execute to get it to work

**Reply by ADavidBailey** [reply]
PS P:\build-scripts> icacls makeFiltered.cmd
makeFiltered.cmd  No permissions are set. All users have full control.
Successfully processed 1 files; Failed processing 0 files
PS P:\build-scripts> icacls P:\py\wExtract.py
P:\py\wExtract.py  No permissions are set. All users have full control.
Successfully processed 1 files; Failed processing 0 files

**Reply by ADavidBailey** [reply]
Removing the # comments fixed the issue.  Thanks.

**Reply by ADavidBailey** [reply]
> Removing the # comments fixed the issue. Thanks.

@Rick-Wilson how do I do in-line comments in a PowerShell CMD?  I googled and it said "Anything following the # on the same line is considered a comment and is not executed by PowerShell."

**Reply by ADavidBailey** [reply]
> I would use Visual Code

I've been running VSCode on my Mac for a long time.  Now that I'm starting to use PowerShell, I'm willing to try using it on Windows.  I have some questions.

Do I need to clone my GitHub to my Windows machine?  What advice do you have for doing that?

Should I use VSCode on both my Mac and Windows?  Would each of the VSCode instances need their own local clone?  I downloaded VSCode and at startup, one of the options is 'Open a local folder' should I select that and continue?

**Reply by ThorvaldAagaard** [reply]
You should be able to use the same clone for both instances. You will have to install the extensions for both, especially git. There might be an issue with linefeeds, but it is worth trying. I use it both but have two different pc's.

**Reply by ADavidBailey** [reply]
> You will have to install the extensions for both, especially git.
It seems that lots of the extensions I have installed (I may not need them) are not on windows or are named differently.  Here are those I have installed on my Mac.

<img width="509" alt="Screenshot 2024-10-24 at 11 51 49 AM" src="https://github.com/user-attachments/assets/ea186fc8-7c6d-46ab-bc4d-2552e400e431">
<img width="445" alt="Screenshot 2024-10-24 at 11 51 25 AM" src="https://github.com/user-attachments/assets/c512bbfe-8def-4259-9408-6bd31724ef99">

Installed...
Codeium
PowerShell Tools for Visual Studio 2022

I see several that start with 'git' but nothing looks familiar

I can't find anything... 
'vs-', 'py', 'remo'

**Reply by ThorvaldAagaard** [reply]
Don't try to sync'em
Just install if needed
You can install git on windows from https://git-scm.com/book/en/v2/Getting-Started-Installing-Git (That is what I gig)

**Reply by ADavidBailey** [reply]
That worked.
I have some new folders:  .git, .vs

Otherwise, it looks like VSCode on my Mac.

**Reply by ADavidBailey** [reply]
Well, I take it all back.  I see the folders; but, they're all empty.  No files.  Then, they're NOT empty.  Files are there.  I don't know what's going on.

I think I might have done two VSCode installs on Windows.

**Reply by ADavidBailey** [reply]
I am hopelessly lost.

**Reply by ThorvaldAagaard** [reply]
Just make sure you open the right folder. Send some screenshots, perhaps younger eyes might help.

**Reply by ADavidBailey** [reply]
Icon on windows desktop
![Screenshot 2024-10-24 at 3 38 50 PM](https://github.com/user-attachments/assets/39d6b3fd-0be2-41ce-ab7b-4ab40fcc432f)
right-click and open 
![Screenshot 2024-10-24 at 4 03 36 PM](https://github.com/user-attachments/assets/2f780730-3eec-4c25-862f-11358f3fc31a)
click open file
![Screenshot 2024-10-24 at 4 04 57 PM](https://github.com/user-attachments/assets/92d53900-a716-4a52-954a-229f28d51bda)
If I navigate from here I can move back to P: and open other folders.  Everything seems fine.  I close and click Open Folder...
![Screenshot 2024-10-24 at 4 07 08 PM](https://github.com/user-attachments/assets/b7e772df-5989-4836-85b2-fb5056467844)
It looks fine; BUT all of the folders are empty!!!!!!!!!!!!!  Like this...
![Screenshot 2024-10-24 at 4 08 22 PM](https://github.com/user-attachments/assets/b71d86c4-a847-42f0-9d4d-767d505bb189)

Toyota!!  as in You asked for it, you got it!

**Reply by ThorvaldAagaard** [reply]
The second image is the bbsa folder, the first is the root folder. Aples and oranges :-)

**Reply by ADavidBailey** [reply]
I did NOT make any effort to open a particular folder.  When one's empty they're all empty -- at least, I never found one that wasn't empty; and I tried many.

**Reply by ThorvaldAagaard** [reply]
If you click open folder you will never see any files (At least not on windows)
Select the folder to open and all the files should be visible in vscode

**Reply by ADavidBailey** [reply]
> Select the folder to open

How do I do that?  The only way I've fount is to click 'Open File...'.  That's weird!  Oh well, I think I understand how to make it work.  Thanks.

![Screenshot 2024-10-24 at 5 32 20 PM](https://github.com/user-attachments/assets/2910284c-dd67-4f5b-884c-2faef55af1eb)

**Comment by ADavidBailey** [comment]
Do both share the same clone -- local files?

On Thu, Oct 24, 2024 at 10:14 AM Thorvald Aagaard ***@***.***>
wrote:

> You should be able to use the same clone for both instances. You will have
> to install the extensions for both, especially git. There might be an issue
> with linefeeds, but it is worth trying. I use it both but have two
> different pc's.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ADavidBailey/Practice-Bidding-Scenarios/discussions/46#discussioncomment-11042505>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BC4TXI3E7UKOKDTQEH43HMLZ5EFEFAVCNFSM6AAAAABQQ325YOVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMBUGI2TANI>
> .
> You are receiving this because you authored the thread.Message ID:
> <ADavidBailey/Practice-Bidding-Scenarios/repo-discussions/46/comments/11042505
> @github.com>
>


-- 
David Bailey
918-520-9700 mobile

**Reply by ThorvaldAagaard** [reply]
If you just open the same folder they do. git information is read from the directory

**Comment by ADavidBailey** [comment]
I think I ready to try this.  I'm not changing the Dealer seed; so, most of the pbn files will have inconsequential changes.

I think I'll start this in the morning.  For some of the steps, the new windows-based approach is slower.  Hopefully, the steps that take a long time -- running Dealer and running BBA -- won't be significantly impacted.

There's no need to rush this; so, let me know if you want to delay it to install changes before I try it.

**Reply by ThorvaldAagaard** [reply]
Just go ahead :-)

**Reply by ADavidBailey** [reply]
I need to increase my git buffer size before I do this.  I would like my rescue team (that's you) to be available a couple of hours after I start it.

**Reply by ADavidBailey** [reply]
http.postbuffer=157286400 is done.

**Reply by ThorvaldAagaard** [reply]
3 o'clock here. I hope you manage zzzzz

**Comment by ADavidBailey** [comment]
That's when I got up this morning.  It's 8:00 PM here.


On Thu, Oct 24, 2024 at 8:00 PM Thorvald Aagaard ***@***.***>
wrote:

> 3 o'clock here. I hope you manage zzzzz
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ADavidBailey/Practice-Bidding-Scenarios/discussions/46#discussioncomment-11046762>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BC4TXI7VZ6ZQTKELM54YHJ3Z5GJ27AVCNFSM6AAAAABQQ325YOVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMBUGY3TMMQ>
> .
> You are receiving this because you authored the thread.Message ID:
> <ADavidBailey/Practice-Bidding-Scenarios/repo-discussions/46/comments/11046762
> @github.com>
>


-- 
David Bailey
918-520-9700 mobile

**Comment by ADavidBailey** [comment]
For testing, I would like to process only a few files in each folder.  I tried setting a global variable, $global:nFiles = 2, and referencing it in each of the ps1 files, $($global:nFiles).

It didn't work.  I suspect because they're NOT "within the same PowerShell session".  It was set in aScriptToRuleThemAll.cmd  (NOT .ps1).   And Referenced in the various makeAll... .ps1 files.

I tried just changing the .cmd to .ps1.  No Joy!!

I tested writing a aaaTest.ps1 file which calls another ps1 file.  That will work.

**Reply by ThorvaldAagaard** [reply]
I have written very few powershell-scripts, so I can't help

**Comment by Rick-Wilson** [comment]
The powershells should probably all take a file pattern input.    You could
then test with an input to the first script like 1*.pbn.  Then it would
just do the first few (the ones where the filename starts with a "1").

**Reply by ADavidBailey** [reply]
If the global variable will work, I'll just have to change the 'master' script to toggle between a test run and a production run.

**Comment by ADavidBailey** [comment]
This morin' I woke up thinkin'...

I'm doin' this wrong.  My mama taught me "All for one!"  I. need a script that does all for 1, then a simple script the loops through the scenarios and executes "All for one".  Just a little restructure project.  Well, maybe not so little.

And then I thought, can I do this in python?  Rick's PowerScript program calls P:/Dealer..., P:/BBA..., and P:/Bridge Composer.  PowerScript is slow; very slow compared to Python.  Can I do all of this in Python?

OneScriptToRuleThemAll loops through the scenarios and calls AllForOne which calls...
- calls Extract to get the dealer code
- calls Dealer to make a pbn
- calls commentStats to comment the Stats and appends them to stats.txt
- calls Bridge Composer to rotate the pbn
- calls Bridge Composer to create a rotated lin
- calls BBA to bid the deal and create bba
- calls filter to create filtered and filtered-out and adds the filter stat to the spreadsheet
- calls setTitle to add the title to the pbn
- calls makeBiddingSheet...

Then, each time I create or update a scenario, I call oneForAll with the name of the scenario.  And, when something big happens, I run OneScriptToRuleThemAll.

**Reply by ThorvaldAagaard** [reply]
Should be possible. There might be some problems awaiting the response from a step

ChatGPT is your freind

**Comment by ADavidBailey** [comment]
I started to ask "can I run a dos command from python and resume when the
command finishes"; but, I asked chrome search, instead:

Yes, you can run a DOS command from Python and resume your Python script
when the command finishes. Here's how you can do it using the subprocess
 module:
Python

import subprocess

# Run the DOS command and wait for it to finish
subprocess.run(["dir", "/w"], shell=True)

# Continue with your Python script
print("DOS command finished. Resuming Python script...")

Explanation:

   - subprocess.run():
   This function runs the specified command. The shell=True argument allows
   you to run the command as if you were typing it in the DOS prompt.
   - ["dir", "/w"]:
   This is a list containing the command and its arguments. In this
   example, it runs the dir command with the w option to display the output
   in wide format.
   - *Waiting for the command to finish:*
   The subprocess.run() function waits for the command to complete before
   returning control to the Python script.



On Sat, Oct 26, 2024 at 5:48 AM Thorvald Aagaard ***@***.***>
wrote:

> Should be possible. There might be some problems awaiting the response
> from a step
>
> ChatGPT is your freind
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ADavidBailey/Practice-Bidding-Scenarios/discussions/46#discussioncomment-11059724>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BC4TXI7DDNQU7QDRMIWWAXDZ5NXQRAVCNFSM6AAAAABQQ325YOVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMBVHE3TENA>
> .
> You are receiving this because you authored the thread.Message ID:
> <ADavidBailey/Practice-Bidding-Scenarios/repo-discussions/46/comments/11059724
> @github.com>
>


-- 
David Bailey
918-520-9700 mobile

**Comment by ADavidBailey** [comment]
@Rick-Wilson came up with a fantastic design for OneScriptToRuleThemAll.  And a way to shorten the name when it's invoked.  I am ecstatic!  I'll let Rick tell you all about it when he's ready.
